instance/models/utils.py

```diff
@@ -44,3 +55,263 @@ def save(self, *args, **kwargs):
         """Call :meth:`full_clean` before saving."""
         self.full_clean()
         super(ValidateModelMixin, self).save(*args, **kwargs)
+
+
+class ClassProperty(property):
+    """ Same as built-in 'property' global but also works when accessed as a class attribute """
+    def __get__(self, cls, owner):
+        return self.fget.__get__(None, owner)()  # pylint: disable=no-member
+
+
+class ResourceState:
+    """
+    A [finite state machine] state class representing the overall state of a resource,
+    such as a server, a database, a DNS entry, etc.
+    """
+    # state_id: A string uniquely identifying this state
+    state_id = None
+
+    @ClassProperty
+    @classmethod
+    def name(cls):
+        """
+        name: Human-readable name of this state, suitable for display as "Status: _______"
+
+        Doesn't need to be a property; override in the subclass using just 'name = "Name"'
+        """
+        return cls.__name__
+
+    @ClassProperty
+    @classmethod
+    def description(cls):
+        """
+        description: Human-readable explanation of this state (1-2 sentences)
+
+        Defaults to the state class's docstring, but you can override it.
+        Doesn't need to be a property; override in the subclass using just 'description = "..."'
+        """
+        return cls.__doc__.strip()
+
+    # Constants:
+    PROGRESS_UNKNOWN = -1
+
+    def __str__(self):
+        return "{} [{}]".format(self.name, self.state_id)
+
+    def __eq__(self, obj):
+        """
+        Syntactic sugar to make comparing states easier.
+
+        obj should be a ResourceState class or instance.
+        """
+        try:
+            return issubclass(self.__class__, obj)
+        except TypeError:  # The provided argument is not a class
+            return type(obj) is type(self)
```

Comment:
I rewrote this a couple times and changed my mind each time about the semantics I wanted. I think the most sensible thing for this semantic sugar is strict type checking, rather than `issubclass` or `isinstance` (just use those explicitly if that's what you want, rather than using `==`), so I've changed the method to that and expanded the tests accordingly.